File: src/state/gitignore.ts
/**
 * `.gitignore` management for E3-owned artifacts (R15, R23a `.backup/`).
 *
 * After `c3p init` the project root's `.gitignore` must list
 * `.claude/` and `.claude-profiles/.meta/` (the umbrella for state, lock,
 * tmp, pending, prior, backup). The init flow lives in E6 but the WRITER
 * lives here so init and any other caller (e.g. recovery diagnostic that
 * detects missing entries) share a single implementation.
 *
 * Idempotent: if the entries already exist (with or without trailing
 * comments, leading whitespace) we don't add duplicates. We append, never
 * rewrite — preserving user-managed gitignore content.
 */

import { promises as fs } from "node:fs";

import { atomicWriteFile, uniqueAtomicTmpPath } from "./atomic.js";
import type { StatePaths } from "./paths.js";

/**
 * Required gitignore entries for E3 artifacts. Order is the order they're
 * appended on first write, with a leading section header so users can see
 * what added them.
 */
export const E3_GITIGNORE_ENTRIES = [
  ".claude/",
  ".claude-profiles/.meta/",
] as const;

const SECTION_HEADER = "# Added by c3p";

/**
 * Result of an idempotent gitignore update. `added` lists the entries we
 * appended; empty array means everything was already present.
 */
export interface GitignoreUpdate {
  added: string[];
  /** True if the file did not exist before this call. */
  created: boolean;
}

/**
 * Ensure all E3 entries appear in `.gitignore`. Creates the file if missing.
 * Appends only the entries that aren't already listed (matched by exact-line
 * equality after trim — comments and indented variants don't count as
 * matches and would be flagged as missing).
 */
export async function ensureGitignoreEntries(paths: StatePaths): Promise<GitignoreUpdate> {
  let existing: string;
  let created = false;
  try {
    existing = await fs.readFile(paths.gitignoreFile, "utf8");
  } catch (err: unknown) {
    if ((err as NodeJS.ErrnoException).code === "ENOENT") {
      existing = "";
      created = true;
    } else {
      throw err;
    }
  }

  const present = new Set(
    existing
      .split(/\r?\n/)
      .map((l) => l.trim())
      .filter((l) => l.length > 0 && !l.startsWith("#")),
  );

  const toAdd = E3_GITIGNORE_ENTRIES.filter((e) => !present.has(e));
  if (toAdd.length === 0) {
    return { added: [], created };
  }

  // Append; ensure a single blank-line separator between existing content
  // and our section header so the file stays readable.
  //
  // Atomic write (multi-reviewer P2, Codex #6): use atomicWriteFile so a
  // crash mid-write doesn't truncate the user's `.gitignore`. Consistent
  // with the rest of E3's write discipline.
  //
  // Tmp staging is placed inside `.claude-profiles/.meta/tmp/` rather than
  // next to `.gitignore` (Sonnet review #4, Gemini #2): a `.gitignore.tmp` at
  // the project root would be visible in `git status` after a crash, and
  // adding `.gitignore.tmp` to the gitignore is circular (the file we're
  // writing IS the gitignore). Cross-filesystem rename is not a concern in
  // practice — `.claude-profiles/` is a sibling of `.gitignore`, both inside
  // the project.
  await fs.mkdir(paths.tmpDir, { recursive: true });
  const trimmed = existing.replace(/\s+$/, "");
  const sep = trimmed.length === 0 ? "" : "\n\n";
  const block = `${sep}${SECTION_HEADER}\n${toAdd.join("\n")}\n`;
  const next = trimmed + block;
  const tmpPath = uniqueAtomicTmpPath(paths.tmpDir, paths.gitignoreFile);
  try {
    await atomicWriteFile(paths.gitignoreFile, tmpPath, next);
  } catch (err) {
    await fs.unlink(tmpPath).catch(() => undefined);
    throw err;
  }
  return { added: [...toAdd], created };
}

File: internal/state/gitignore.go
package state

import (
	"errors"
	"os"
	"strings"
)

// GitignoreEntries are the project-root .gitignore lines required by D5
// artifacts (R15, R23a/R28). Order is the order they're appended on first
// write; a section header precedes them so users see what added them.
//
// CLAUDE.md.*.tmp is the splice-write staging file; applyRootSplice deliberately
// places it next to the live CLAUDE.md to keep the temp+rename on the same
// filesystem (no EXDEV risk). sweepRootClaudeMdTmps removes them on the next
// c3p op, but a crash followed by `git add -A` could otherwise stage the debris.
var GitignoreEntries = []string{
	".claude/",
	".claude-profiles/.meta/",
	"CLAUDE.md.*.tmp",
}

const gitignoreSectionHeader = "# Added by c3p"

// GitignoreUpdate is the result of EnsureGitignoreEntries. Added is the
// entries we appended; an empty slice means everything was already present.
// Created is true iff the file did not exist before this call.
type GitignoreUpdate struct {
	Added   []string
	Created bool
}

// EnsureGitignoreEntries appends each missing GitignoreEntries entry to the
// project-root .gitignore (R15). Idempotent: existing entries (matched by
// exact-line equality after trim, comments and indented variants do NOT
// count as matches and would be flagged as missing) are left in place.
//
// Atomic write: tmp staging is placed inside .claude-profiles/.meta/tmp/
// (NOT next to .gitignore at the project root). A .gitignore.tmp at the
// project root would be visible in `git status` after a crash, and adding
// .gitignore.tmp to the gitignore is circular (the file we're writing IS
// the gitignore). Cross-filesystem rename is not a concern in practice —
// .claude-profiles/ is a sibling of .gitignore, both inside the project.
func EnsureGitignoreEntries(paths StatePaths) (GitignoreUpdate, error) {
	existingBytes, err := os.ReadFile(paths.GitignoreFile)
	created := false
	existing := ""
	if err != nil {
		if !errors.Is(err, os.ErrNotExist) {
			return GitignoreUpdate{}, err
		}
		created = true
	} else {
		existing = string(existingBytes)
	}

	present := map[string]struct{}{}
	for _, line := range strings.Split(existing, "\n") {
		// Mirror the TS reference's split(/\r?\n/) by trimming a CR before
		// the trim-and-empty check.
		trimmed := strings.TrimRight(line, "\r")
		trimmed = strings.TrimSpace(trimmed)
		if trimmed == "" {
			continue
		}
		if strings.HasPrefix(trimmed, "#") {
			continue
		}
		present[trimmed] = struct{}{}
	}

	toAdd := make([]string, 0, len(GitignoreEntries))
	for _, e := range GitignoreEntries {
		if _, ok := present[e]; !ok {
			toAdd = append(toAdd, e)
		}
	}
	if len(toAdd) == 0 {
		return GitignoreUpdate{Added: []string{}, Created: created}, nil
	}

	// Trim trailing whitespace; insert a single blank-line separator between
	// existing content and our section header so the file stays readable.
	trimmed := strings.TrimRight(existing, " \t\n\r")
	sep := ""
	if trimmed != "" {
		sep = "\n\n"
	}
	block := sep + gitignoreSectionHeader + "\n" + strings.Join(toAdd, "\n") + "\n"
	next := trimmed + block

	if err := os.MkdirAll(paths.TmpDir, 0o755); err != nil {
		return GitignoreUpdate{}, err
	}
	tmpPath := UniqueAtomicTmpPath(paths.TmpDir, paths.GitignoreFile)
	if err := AtomicWriteFile(paths.GitignoreFile, tmpPath, []byte(next)); err != nil {
		_ = os.Remove(tmpPath)
		return GitignoreUpdate{}, err
	}
	added := append([]string(nil), toAdd...)
	return GitignoreUpdate{Added: added, Created: created}, nil
}
